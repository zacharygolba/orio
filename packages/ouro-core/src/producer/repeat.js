// @flow

import * as result from 'ouro-result'
import { AsIterator, ToString } from 'ouro-traits'

import type { Producer } from '../types'

@ToString
@AsIterator
export default class Repeat<T> implements Producer<T> {
  /*:: @@iterator: () => Iterator<T> */
  done: boolean
  value: T

  constructor(value: T) {
    this.done = false
    this.value = value
  }

  drop(): void {
    this.done = true
  }

  next(): IteratorResult<T, void> {
    if (this.done) {
      return result.done()
    }

    return result.next(this.value)
  }
}
