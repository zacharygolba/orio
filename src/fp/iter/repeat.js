// @flow

export default function* repeat<T>(value: T): Iterator<T> {
  for (;;) {
    yield value
  }
}
