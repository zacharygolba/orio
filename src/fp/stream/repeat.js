// @flow

export default async function* repeat<T>(value: T): AsyncIterator<T> {
  for (;;) {
    yield value
  }
}
