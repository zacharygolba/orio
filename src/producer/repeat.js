// @flow

import * as result from '../result'

import ProducerBase from './base'

export default class RepeatProducer<T> extends ProducerBase<T> {
  value: T

  constructor(value: T) {
    super()
    this.value = value
  }

  next(): IteratorResult<T, void> {
    return result.next(this.value)
  }

  sizeHint(): number {
    return Infinity
  }
}
