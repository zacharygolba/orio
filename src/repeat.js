// @flow

import { impl } from './iter'
import type { Iter } from './iter'

export default function repeat<T: any>(value: T): Iter<T> {
  return new Repeat(value)
}

const Repeat = impl(class RepeatIterator<T> {
  value: T

  constructor(value: T) {
    this.value = value
  }

  next(): IteratorResult<T, void> {
    return {
      done: false,
      value: this.value,
    }
  }

  sizeHint(): number {
    return Infinity
  }
})
