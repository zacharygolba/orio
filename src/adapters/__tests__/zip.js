// @flow

import Zip from '../zip'
import intoIter from '../../utils/into-iter'
import repeat from '../../repeat'

test('#next()', () => {
  {
    const iter = new Zip(repeat('test'), intoIter([1, 2, 3]))

    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
  }

  {
    const iter = new Zip(intoIter([1, 2, 3]), repeat('test'))

    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
  }
})

test('#sizeHint()', () => {
  {
    const iter = new Zip(repeat('test'), intoIter([1, 2, 3]))
    expect(iter.sizeHint()).toEqual(3)
  }

  {
    const iter = new Zip(intoIter([1, 2, 3]), repeat('test'))
    expect(iter.sizeHint()).toEqual(3)
  }

  {
    const iter = new Zip(repeat('test'), repeat('test'))
    expect(iter.sizeHint()).toEqual(Infinity)
  }
})
