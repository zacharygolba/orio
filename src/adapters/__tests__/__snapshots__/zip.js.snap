// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#next() 1`] = `
Object {
  "done": false,
  "value": Array [
    "test",
    1,
  ],
}
`;

exports[`#next() 2`] = `
Object {
  "done": false,
  "value": Array [
    "test",
    2,
  ],
}
`;

exports[`#next() 3`] = `
Object {
  "done": false,
  "value": Array [
    "test",
    3,
  ],
}
`;

exports[`#next() 4`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`#next() 5`] = `
Object {
  "done": false,
  "value": Array [
    1,
    "test",
  ],
}
`;

exports[`#next() 6`] = `
Object {
  "done": false,
  "value": Array [
    2,
    "test",
  ],
}
`;

exports[`#next() 7`] = `
Object {
  "done": false,
  "value": Array [
    3,
    "test",
  ],
}
`;

exports[`#next() 8`] = `
Object {
  "done": true,
  "value": undefined,
}
`;
